oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/InMemoryOAuth2AuthorizationService.java

```diff
@@ -108,4 +113,26 @@ public int hashCode() {
 			return Objects.hash(this.registeredClientId, this.principalName);
 		}
 	}
+
+	@Override
+	public Optional<OAuth2Authorization> findByTokenWithHint(String token,
```

Comment:
One quick note here. I realize the existing _findToken_ method will retrieve null if the token doesn't match the provided _tokenType_. However, here we want to make the search extensive to any supported token type, as per the specs:

> token_type_hint
>       OPTIONAL.  A hint about the type of the token submitted for
>       introspection.  The protected resource MAY pass this parameter to
>       help the authorization server optimize the token lookup.  **If the
>       server is unable to locate the token using the given hint, it MUST
>       extend its search across all of its supported token types.**  An
>       authorization server MAY ignore this parameter, particularly if it
>       is able to detect the token type automatically.  Values for this
>       field are defined in the "OAuth Token Type Hints" registry defined
>       in OAuth Token Revocation [RFC7009].

	  
I realized the Token Revocation code also uses this logic, but the specs indicate the same as above, so IMO that should be changed to use this method instead. If you agree, I'll create an issue for that :)
Finally, I know https://github.com/spring-projects-experimental/spring-authorization-server/pull/162 is already closed and I'm ok with that decision, but I caught this issue using those tests locally (in our current Integration tests we're mocking up these interactions), just to point out that they might be suitable for some edge cases :)

Reply:
Note: I now created https://github.com/spring-projects-experimental/spring-authorization-server/issues/175 to fix this in the Token Revokation logic.